### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by jesse-graffle**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,44 @@
 # flow-event-processor
 A processor to monitor events in the flow blockchain
+
+Configuration:
+- FlowNode: MainNet or TestNet
```
Comment (dakota-graffle) [inline README.md]: This work with emulator?

File: Graffle.FlowEventProcessor/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,14 @@
+FROM mcr.microsoft.com/dotnet/sdk:5.0 AS build
```
Comment (dakota-graffle) [inline Graffle.FlowEventProcessor/Dockerfile]: This is the prod dockerfile lets make sure we optimize size with alpines

File: Graffle.FlowEventProcessor/Program.cs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                                    EventDate = @event.BlockTimestamp,
+                                    FlowBlockId = @event.BlockIdHash,
+                                    FlowEventId = eventId,
+                                    FlowTransactionId = @event.TransactionId.ToHash()
+                                };
+
+                                //Send webhook
+                                var jsonMesage = System.Text.Json.JsonSerializer.Serialize(flowEvent);
+                                using (var content = new StringContent(jsonMesage)) {
+                                    using (var client = new HttpClient()) {
```
Comment (dakota-graffle) [inline Graffle.FlowEventProcessor/Program.cs]: I would create a HttpClient singleton at start up. Going to get pretty wasteful on high throughput projects

**Review by dakota-graffle — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (jesse-graffle) [reply README.md]: It will currently work with the emulator when running from VS Code, but not from docker-compose yet.

File: Graffle.FlowEventProcessor/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,16 @@
+FROM mcr.microsoft.com/dotnet/sdk:5.0-alpine AS build
```
Comment (dakota-graffle) [inline Graffle.FlowEventProcessor/Dockerfile:1]: This file needs to be updated to reflect prod. This is what will be used in the build pipeline

File: Graffle.FlowEventProcessor/Dockerfile.compose [hunk 5]
```diff
@@ -0,0 +1,8 @@
+FROM mcr.microsoft.com/dotnet/sdk:5.0 AS build
```
Comment (dakota-graffle) [inline Graffle.FlowEventProcessor/Dockerfile.compose]: Is this file needed?

File: Graffle.FlowEventProcessor/Dockerfile.development [hunk 6]
```diff
@@ -0,0 +1,19 @@
+FROM mcr.microsoft.com/dotnet/sdk:5.0-alpine AS build
```
Comment (dakota-graffle) [inline Graffle.FlowEventProcessor/Dockerfile.development]: This can either be removed or we need to switch back to using the full SDK and runtimes so people can debug. This current state should be in the Dockerfile

File: Graffle.FlowEventProcessor/Program.cs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            var config = new ConfigurationBuilder()
+            .SetBasePath(AppDomain.CurrentDomain.BaseDirectory)
+            .AddJsonFile("appsettings.json", true)
+            .AddJsonFile($"appsettings.{environmentName}.json", optional: true)
+            .AddEnvironmentVariables()
+            .Build();
+
+            Console.WriteLine($"Loaded Settings Files.");
+
+            var nodeName = config.GetValue<string>("FlowNode") ?? "MainNet";
```
Comment (dakota-graffle) [inline Graffle.FlowEventProcessor/Program.cs]: Lets not default to any node and just throw and error and tell them to pass something in

Comment (dakota-graffle) [reply Graffle.FlowEventProcessor/Program.cs]: Maybe check if its a valid selection of Test,Main, or Emulator(When thats ready)

File: Graffle.FlowEventProcessor/Program.cs [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+            var maximumBlockScanRange = config.GetValue<ulong>("MaximumBlockScanRange");
```
Comment (dakota-graffle) [inline Graffle.FlowEventProcessor/Program.cs]: Can this be set as ulong? and have a check if it has a value down below on line 49?

File: Graffle.FlowEventProcessor/Program.cs [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+            if(string.IsNullOrWhiteSpace(webhookUrl)) {
+                throw new Exception("Specify WebhookUrl environment variable.");
+            }
+
+            if(string.IsNullOrWhiteSpace(eventId)) {
+                throw new Exception("Specify EventId environment variable.");
+            }
+            
+            if(maximumBlockScanRange == 0){
```
Comment (dakota-graffle) [inline Graffle.FlowEventProcessor/Program.cs]: As mention above lets check .HasValue as well as if its <= 0 so they cant pass in negative numbers

File: Graffle.FlowEventProcessor/Program.cs [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                                    FlowTransactionId = @event.TransactionId.ToHash()
+                                };
+
+                                //Send webhook
+                                var jsonMesage = System.Text.Json.JsonSerializer.Serialize(flowEvent);
+                                using (var content = new StringContent(jsonMesage)) {
+                                    if(verbose) {
+                                        Console.WriteLine($"Posting event to {webhookUrl}: {Environment.NewLine}     {jsonMesage}");
+                                    }
+                                    await httpClient.PostAsync(webhookUrl, content);
```
Comment (dakota-graffle) [inline Graffle.FlowEventProcessor/Program.cs]: Lets add another 3 try Polly on sending out the webhook

File: README.md [hunk 11]
```diff
@@ -1,2 +1,65 @@
...
+Create a .env.local file in the root of your repository, next to docker-compose.yml:
+```
+WEBHOOKURL=https://<your url>
+EVENTID=A.c1e4f4f4c4257510.Market.MomentPurchased
+VERBOSE=true
+```
+
+Open a terminal window in the root of your repository and run one of the following:
+```
+ docker-compose --env-file .\.env.local build test_net_graffle_flow_event_processor
```
Comment (dakota-graffle) [inline README.md]: is env file needed on build?

File: docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    environment:
+      - ASPNETCORE_ENVIRONMENT=local
+      - FlowNode=MainNet
+      - MaximumBlockScanRange=200
+      - WebhookUrl=${WEBHOOKURL}
+      - EventId=${EVENTID}
+      - Verbose=${VERBOSE}
+    image: 'graffle/flow_event_processor:dev'
+    volumes:
+      - ./Graffle.FlowEventProcessor/bin/Debug/net5.0:/app
```
Comment (dakota-graffle) [inline docker-compose.yml:31]: if we are pointing a volume at the debug folder than the Docker.Development file doesnt need to copy all the files over as the container itself has direct access to the build files on the machine.

**Review by dakota-graffle — CHANGES_REQUESTED** [review]

**Review by dakota-graffle — APPROVED** [review]
